src/python/pants/option/parser.py

```diff
@@ -315,84 +326,101 @@ def add_flag_val(v: Optional[Union[int, float, bool, str]]) -> None:
+                    # name, display it. This will also suggest long-form options such as --verbose
+                    # for an attempted -v (if -v isn't defined as an option).
+                    substring_matching_flags.append(other_complete_flag_name)
                 else:
-                    # If an unscoped option name is similar to the unscoped option from the command line
-                    # according to --option-name-check-distance, display the matching scoped option name. This
-                    # covers misspellings!
-                    unscoped_option_levenshtein_distance = Levenshtein.distance(
+                    # If an unscoped option name is similar to the unscoped option from the command
+                    # line according to --option-name-check-distance, display the matching scoped
+                    # option name. This covers misspellings.
+                    unscoped_option_edit_distance: int = Levenshtein.distance(
                         flag_normalized_unscoped_name, other_normalized_unscoped_name
                     )
-                    if unscoped_option_levenshtein_distance <= levenshtein_max_distance:
-                        # NB: We order the matched flags by Levenshtein distance compared to the entire option string!
-                        fully_scoped_levenshtein_distance = Levenshtein.distance(
+                    if unscoped_option_edit_distance <= max_edit_distance:
+                        # NB: We order the matched flags by Levenshtein distance compared to the
+                        # entire option string.
+                        fully_scoped_edit_distance: int = Levenshtein.distance(
                             flag_normalized_scoped_name, other_normalized_scoped_name
                         )
-                        levenshtein_matching_option_names[fully_scoped_levenshtein_distance].append(
+                        edit_distance_to_matching_flags[fully_scoped_edit_distance].append(
                             other_complete_flag_name
                         )
 
             # If any option name matched or started with the invalid flag in any scope, put that
-            # first. Then, display the option names matching in order of overall edit distance, in a deterministic way.
-            all_matching_scoped_option_names = substring_matching_option_names + [
-                flag
-                for distance in sorted(levenshtein_matching_option_names.keys())
-                for flag in sorted(levenshtein_matching_option_names[distance])
+            # first. Then, display the option names matching in order of overall edit distance.
+            # Flags with the same distance will be sorted alphabetically.
+            all_matching_scoped_flags = [
+                *substring_matching_flags,
+                *itertools.chain.from_iterable(
+                    matched_flags
+                    for _, matched_flags in sorted(edit_distance_to_matching_flags.items())
```

Comment:
Tuples are sorted by first comparing the first element, then the second element, and so on. So, this will first sort by edit distance, then by flag name in alphabetical order.